### PR TITLE
Uml 900 track form validation errors in GA

### PR DESCRIPTION
**Pull request by GemTay**
# Purpose

Enables tracking of all form error messages in google analytics to help with user research and improvements

Fixes UML-900

## Approach

Created a new function in the googleAnalytics.js class to select all field attributes with the class `govuk-form-group--error`. Following the structure of the event action, category and label detailed in the ticket, for each field error, it extracts the field label eg "Enter your email address", input id eg "email", and error messages * to be used as the components of the event. Each field can have more than one error message, for example the various criteria when creating a new password. Therefore a second loop is performed on each of these error messages to ensure that an event is fired for each one. 

A Unit test has been created and the events have been manually tested using the google analytics debugger Chrome extension. Heres an example of the events being triggered on the create account page:
image

## Checklist

* [x] I have performed a self-review of my own code
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [x] I have added tests to prove my work
* [ ] I have added welsh translation tags and updated translation files
* [ ] I have run an accessibility tool on any pages I have made changes to and fixed any issues found
* [ ] The product team have tested these changes

### CODE REVIEW

File: service-front/app/src/Common/templates/partials/govuk_form.html.twig [hunk 1]
```diff
@@ -71,15 +71,14 @@
...
+        {{ govuk_error(element) }}
         <div class="govuk-checkboxes">
             <div class="govuk-checkboxes__item">
 
                 <input class="govuk-checkboxes__input" id="{{ element.getName() }}" name="{{ element.getName() }}" value="{{ element.getCheckedValue() }}" type="checkbox" {% if element.getValue() == element.getCheckedValue() %}checked="checked"{% endif %}>
 
                 <label class="govuk-label govuk-checkboxes__label" for="{{ element.getName() }}">
                     {{ label | raw }}
-
-                    {{ govuk_error(element) }}
```
Comment (GemTay) [inline service-front/app/src/Common/templates/partials/govuk_form.html.twig:82]: The element was moved up as before it was in this odd position:
image

Comment (GemTay) [reply service-front/app/src/Common/templates/partials/govuk_form.html.twig:82]: Now:
image

File: service-front/app/src/Common/templates/partials/govuk_form.html.twig [hunk 2]
```diff
@@ -96,6 +95,7 @@
     {%- endfor -%}
 
     <div class="govuk-form-group {% if element.getMessages() is not empty %}govuk-form-group--error{% endif %}">
+        {{ govuk_error(element) }}
```
Comment (GemTay) [inline service-front/app/src/Common/templates/partials/govuk_form.html.twig:98]: the error element was missing from the radio block, so it only appeared in the error summary

File: service-front/web/src/javascript/googleAnalytics.js [hunk 3]
```diff
@@ -75,4 +76,19 @@ export default class GoogleAnalytics {
...
+    {
+        let errorFields = document.getElementsByClassName('govuk-form-group--error');
+        for (let i = 0, len = errorFields.length; i < len; i++) {
+            let fieldLabel = (errorFields[i].getElementsByTagName('label')[0].textContent).trim();
+            let inputId = (errorFields[i].getElementsByTagName('input')[0].getAttribute('id'));
+            // there can be more than one error message per field eg password rules
+            let errorMessages = (errorFields[i].querySelectorAll('.govuk-error-message'));
+            for (let x = 0, len = errorMessages.length; x < len; x++) {
+                let errorMessage = errorMessages[x].textContent.replace("Error:", "").trim();
+                this.trackEvent(fieldLabel, 'Form errors', ('#id-' + inputId + ' - ' + errorMessage));
```
Comment (mattmachell) [inline service-front/web/src/javascript/googleAnalytics.js:90]: You might want to look at what Digideps did to namespace GA event tracking. https://github.com/ministryofjustice/opg-digideps/blob/main/docs/runbooks/GOOGLE-ANALYTICS.md

Comment (GemTay) [reply service-front/web/src/javascript/googleAnalytics.js:90]: I've roughly followed LPA live's format:
https://opgtransform.atlassian.net/wiki/spaces/LDS/pages/298188834/Event+Tracking

File: service-front/web/src/javascript/googleAnalytics.js [hunk 4]
```diff
@@ -75,4 +76,19 @@ export default class GoogleAnalytics {
             });
         }
     }
+
+    _trackFormValidationErrors()
+    {
+        let errorFields = document.getElementsByClassName('govuk-form-group--error');
+        for (let i = 0, len = errorFields.length; i < len; i++) {
+            let fieldLabel = (errorFields[i].getElementsByTagName('label')[0].textContent).trim();
+            let inputId = (errorFields[i].getElementsByTagName('input')[0].getAttribute('id'));
```
[the remark this answers is not in the document]
Comment (GemTay) [reply service-front/web/src/javascript/googleAnalytics.js:85]: good idea, i like it :)

File: service-front/app/src/Common/templates/partials/govuk_form.html.twig (hunk 1 above)
Comment (cooperaj) [reply service-front/app/src/Common/templates/partials/govuk_form.html.twig:82]: Does the output (with error) look like how GDS would want it to? Is it accessible etc.

File: service-front/app/src/Common/templates/partials/govuk_form.html.twig (hunk 2 above)
Comment (cooperaj) [reply service-front/app/src/Common/templates/partials/govuk_form.html.twig:98]: I guess the same applies here.

**Review by cooperaj — DISMISSED** [review]
Looks solid.

File: service-front/app/src/Common/templates/partials/govuk_form.html.twig (hunk 1 above)
Comment (GemTay) [reply service-front/app/src/Common/templates/partials/govuk_form.html.twig:82]: The output looks the same as the GDS template, you can see an example with the error message near the bottom: https://design-system.service.gov.uk/components/radios/ 

Good spot, i've just made some tweaks to fix some accessibility issues for the radio buttons. Check boxes were fine accessibility wise :)

**Review by cooperaj — APPROVED** [review]
LGTM

**Pull request merged**